File: app/src/main/java/org/schabi/newpipe/streams/SrtFromTtmlWriter.java
package org.schabi.newpipe.streams;

import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.nodes.Node;
import org.jsoup.nodes.TextNode;
import org.jsoup.parser.Parser;
import org.jsoup.select.Elements;
import org.schabi.newpipe.streams.io.SharpStream;

import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.nio.charset.Charset;
import java.nio.charset.StandardCharsets;

/**
 * @author kapodamy
 */
public class SrtFromTtmlWriter {
    private static final String NEW_LINE = "\r\n";

    private final SharpStream out;
    private final boolean ignoreEmptyFrames;
    private final Charset charset = StandardCharsets.UTF_8;

    private int frameIndex = 0;

    public SrtFromTtmlWriter(final SharpStream out, final boolean ignoreEmptyFrames) {
        this.out = out;
        this.ignoreEmptyFrames = ignoreEmptyFrames;
    }

    private static String getTimestamp(final Element frame, final String attr) {
        return frame
                .attr(attr)
                .replace('.', ','); // SRT subtitles uses comma as decimal separator
    }

    private static void extractTextFromNode(final Node node, final StringBuilder text) {
        if (node instanceof TextNode) {
            text.append(((TextNode) node).text());
        } else if (node instanceof Element) {
            final Element element = (Element) node;
            if (element.tagName().equalsIgnoreCase("br")) {
                text.append(NEW_LINE);
            } else {
                // Recursively process child nodes (handles nested span elements)
                for (final Node child : element.childNodes()) {
                    extractTextFromNode(child, text);
                }
            }
        }
    }

    private void writeFrame(final String begin, final String end, final StringBuilder text)
            throws IOException {
        writeString(String.valueOf(frameIndex++));
        writeString(NEW_LINE);
        writeString(begin);
        writeString(" --> ");
        writeString(end);
        writeString(NEW_LINE);
        writeString(text.toString());
        writeString(NEW_LINE);
        writeString(NEW_LINE);
    }

    private void writeString(final String text) throws IOException {
        out.write(text.getBytes(charset));
    }

    public static String convertTtmlToSrt(final String ttmlContent) {
        final StringBuilder srtOutput = new StringBuilder();
        int frameIndex = 0;

        try {
            final Document doc = Jsoup.parse(ttmlContent, "", Parser.xmlParser());
            final StringBuilder text = new StringBuilder(128);
            final Elements paragraphList = doc.select("body > div > p");

            if (paragraphList.size() < 1) {
                return "";
            }

            for (final Element paragraph : paragraphList) {
                text.setLength(0);

                for (final Node children : paragraph.childNodes()) {
                    extractTextFromNode(children, text);
                }

                if (text.length() < 1) {
                    continue;
                }

                final String begin = getTimestamp(paragraph, "begin");
                final String end = getTimestamp(paragraph, "end");

                srtOutput.append(++frameIndex);
                srtOutput.append(NEW_LINE);
                srtOutput.append(begin);
                srtOutput.append(" --> ");
                srtOutput.append(end);
                srtOutput.append(NEW_LINE);
                srtOutput.append(text.toString());
                srtOutput.append(NEW_LINE);
                srtOutput.append(NEW_LINE);
            }
        } catch (Exception e) {
            return "";
        }

        return srtOutput.toString();
    }

    public void build(final SharpStream ttml) throws IOException {
        /*
         * TTML parser with BASIC support
         * multiple CUE is not supported
         * styling is not supported
         * tag timestamps (in auto-generated subtitles) are not supported, maybe in the future
         * also TimestampTagOption enum is not applicable
         * Language parsing is not supported
         */

        // parse XML
        final byte[] buffer = new byte[(int) ttml.available()];
        ttml.read(buffer);
        final Document doc = Jsoup.parse(new ByteArrayInputStream(buffer), "UTF-8", "",
                Parser.xmlParser());

        final StringBuilder text = new StringBuilder(128);
        final Elements paragraphList = doc.select("body > div > p");

        // check if has frames
        if (paragraphList.size() < 1) {
            return;
        }

        for (final Element paragraph : paragraphList) {
            text.setLength(0);

            for (final Node children : paragraph.childNodes()) {
                extractTextFromNode(children, text);
            }

            if (ignoreEmptyFrames && text.length() < 1) {
                continue;
            }

            final String begin = getTimestamp(paragraph, "begin");
            final String end = getTimestamp(paragraph, "end");

            writeFrame(begin, end, text);
        }
    }
}

File: app/src/main/java/org/schabi/newpipe/streams/SrtFromTtmlWriter.kt
package org.schabi.newpipe.streams

import java.io.ByteArrayInputStream
import okio.IOException
import java.nio.charset.Charset
import java.nio.charset.StandardCharsets
import com.fleeksoft.ksoup.Ksoup
import com.fleeksoft.ksoup.nodes.Document
import com.fleeksoft.ksoup.nodes.Element
import com.fleeksoft.ksoup.nodes.Node
import com.fleeksoft.ksoup.nodes.TextNode
import com.fleeksoft.ksoup.parser.Parser
import org.schabi.newpipe.streams.io.SharpStream

/**
 * Converts TTML subtitles to SRT format.
 *
 * References:
 *  - TTML 2.0 (W3C): https://www.w3.org/TR/ttml2/
 *  - SRT format: https://en.wikipedia.org/wiki/SubRip
 */
class SrtFromTtmlWriter(private val out: SharpStream, private val ignoreEmptyFrames: Boolean) {
    private val charset: Charset = StandardCharsets.UTF_8

    // According to the SubRip (.srt) specification, subtitle
    // numbering must start from 1.
    // Some players accept 0 or even negative indices,
    // but to ensure compliance we start at 1.
    private var frameIndex = 1

    private fun getTimestamp(frame: Element, attr: String): String {
        return frame.attr(attr).replace('.', ',') // SRT subtitles uses comma as decimal separator
    }

    @Throws(IOException::class)
    private suspend fun writeFrame(begin: String, end: String, text: StringBuilder) {
        writeString(frameIndex.toString())
        frameIndex += 1
        writeString(NEW_LINE)
        writeString(begin)
        writeString(" --> ")
        writeString(end)
        writeString(NEW_LINE)
        writeString(text.toString())
        writeString(NEW_LINE)
        writeString(NEW_LINE)
    }

    @Throws(IOException::class)
    private suspend fun writeString(text: String) {
        out.write(text.toByteArray(charset))
    }

    private fun decodeXmlEntities(encodedEntities: String): String {
        return Parser.unescapeEntities(encodedEntities, true)
    }

    private fun normalizeLineBreakForSrt(text: String): String {
        return text.replace("\r\n", "\n")
            .replace("\r", "\n")
            .replace("\n", NEW_LINE)
    }

    private fun normalizeForSrt(actualText: String): String {
        var cleaned = actualText
            .replace('\u00A0', ' ') // Non-breaking space
            .replace('\u202F', ' ') // Narrow no-break space
            .replace('\u205F', ' ') // Medium mathematical space
            .replace('\u3000', ' ') // Ideographic space
            .replace(Regex("[\\u2000-\\u200A]"), " ") // Whitespace characters

        cleaned = cleaned.replace(Regex("[\\u200B-\\u200F]"), "") // Non-spacing characters

        cleaned = cleaned.replace(Regex("[\\u0000-\\u0008\\u000B\\u000C\\u000E-\\u001F]"), "")

        cleaned = cleaned.replace('\t', ' ')

        return normalizeLineBreakForSrt(cleaned)
    }

    private fun sanitizeFragment(raw: String?): String {
        if (null == raw) return ""
        val actualCharacters = decodeXmlEntities(raw)
        return normalizeForSrt(actualCharacters)
    }

    private fun traverseChildNodesForNestedTags(parent: Node, text: StringBuilder) {
        for (child in parent.childNodes()) {
            extractText(child, text)
        }
    }

    private fun extractText(node: Node, text: StringBuilder) {
        if (node is TextNode) {
            val srtContent = sanitizeFragment(node.getWholeText())
            text.append(srtContent)
        } else if (node is Element) {
            if (node.tagName().equals("br", ignoreCase = true)) {
                text.append(NEW_LINE)
            }
        }
        traverseChildNodesForNestedTags(node, text)
    }

    @Throws(IOException::class)
    suspend fun build(ttml: SharpStream) {
        // parse XML
        val bufferSize = ttml.available().toInt()
        val buffer = ByteArray(bufferSize)
        ttml.read(buffer)
        val doc: Document = Ksoup.parse(
            html = String(buffer, charset),
            parser = Parser.xmlParser(),
            baseUri = ""
        )

        val text = StringBuilder(128)
        val paragraphList = doc.select("body > div > p")

        if (paragraphList.isEmpty()) {
            return
        }

        for (paragraph in paragraphList) {
            text.setLength(0)
            extractText(paragraph, text)

            if (ignoreEmptyFrames && text.isEmpty()) {
                continue
            }

            val begin = getTimestamp(paragraph, "begin")
            val end = getTimestamp(paragraph, "end")
            writeFrame(begin, end, text)
        }
    }

    companion object {
        private const val NEW_LINE = "\r\n"
    }
}
